homework/index.js

```diff
@@ -1,46 +1,99 @@
-      }
-    };
-    xhr.onerror = () => cb(new Error('Network request failed'));
-    xhr.send();
-  }
+function fetchJSON(url, cb) {
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', url);
+  xhr.responseType = 'json';
+  xhr.onload = () => {
+    if (xhr.status >= 200 && xhr.status <= 299) {
+      cb(null, xhr.response);
+    } else {
+      cb(new Error(`Network error:${xhr.status} -${xhr.statusText}`));
+    }
+  };
+  xhr.onerror = () => cb(new Error('Network request failed'));
+  xhr.send();
+}
 
-  function createAndAppend(name, parent, options = {}) {
-    const elem = document.createElement(name);
-    parent.appendChild(elem);
-    Object.keys(options).forEach(key => {
-      const value = options[key];
-      if (key === 'text') {
-        elem.textContent = value;
-      } else {
-        elem.setAttribute(key, value);
-      }
-    });
-    return elem;
-  }
+function createAndAppend(name, parent, options = {}) {
+  const elem = document.createElement(name);
+  parent.appendChild(elem);
+  Object.keys(options).forEach(key => {
+    const value = options[key];
+    if (key === 'text') {
+      elem.textContent = value;
+    } else {
+      elem.setAttribute(key, value);
+    }
+  });
+  return elem;
+}
+
+function main(url) {
+  fetchJSON(url, (err, repositories) => {
+    const root = document.getElementById('root');
+    if (err) {
+      createAndAppend('div', root, { text: err.message, class: 'alert-error' });
+      return;
+    }
+    const repNames = [];
+    createAndAppend('select', root, { id: 'main-select' });
+    createAndAppend('div', root, { class: 'portion-parent', id: 'parent' });
+    const parent = document.getElementById('parent');
+    createAndAppend('div', parent, { class: 'left-portion', id: 'lft' });
+    createAndAppend('div', parent, { class: 'right-portion', id: 'rt' });
 
-  function main(url) {
-    fetchJSON(url, (err, repositories) => {
-      const root = document.getElementById('root');
-      if (err) {
-        createAndAppend('div', root, { text: err.message, class: 'alert-error' });
-        return;
+    const leftPortion = document.getElementById('lft');
+    const rightPortion = document.getElementById('rt');
+    for (let i = 0; i < repositories.length; i++) {
+      const rep = repositories[i];
+      repNames.push(rep.name);
+    }
+    const selectItem = document.getElementById('main-select');
+
+    repNames.sort();
+    for (let i = 0; i < repNames.length; i++) {
+      createAndAppend('option', selectItem, {
+        value: i,
+        text: repNames[i],
+      });
+    }
+    selectItem.selectedIndex = 0;
+    selectItem.addEventListener('change', () => {
+      const reponame = selectItem.options[selectItem.selectedIndex].text;
+      for (let i = 0; i < repositories.length; i++) {
+        if (repositories[i].name === reponame) {
+          const repo = repositories[i];
+          leftPortion.innerHTML = '';
```

Comment:
The HTML structure you generate leaves room for improvement. If it gives you difficulty it will help to draw up a 'plan' on what HTML you want to generate, and then devise the JavaScript that is needed to achieve that. If you try it directly in JavaScript without a plan you have two problems to deal with at the same time. Therefore: _divide and conquer_.

Here is what you generate now (using the `alumni` repo as example):

```html
<div id="root">
  <select id="main-select">
    <option value="0">AngularJS</option>
    <!-- - more options -->
    <option value="46">ukulele</option>
  </select>
  <div class="portion-parent" id="parent">
    <div class="left-portion" id="lft">
      <p>alumni</p>
      <p>Alumni Guides, everything you need to land a job</p>
      <p>9</p>
      <p>2019-06-21T11:38:21Z</p>
    </div>
    <div class="right-portion" id="rt">
      <div class="rightParent" id="contList0">
        <img src="https://avatars0.githubusercontent.com/u/15912395?v=4" />
        <p>gijscor</p>
        <p>47</p>
      </div>
      <div class="rightParent" id="contList1">
        <img src="https://avatars1.githubusercontent.com/u/37488847?v=4" />
        <p>wouterkleijn</p>
        <p>11</p>
      </div>
    </div>
  </div>
</div>
```

This is what I generate in my solution to this homework:

```html
<div id="root">
  <header class="header">
    <p>HYF Repositories</p>
    <select class="repo-selector">
      <option value="0">alumni</option>
      <!-- - more options -->
      <option value="46">ukulele</option>
    </select>
  </header>
  <div id="container">
    <div class="left-div whiteframe">
      <table>
        <tbody>
          <tr>
            <td class="label">Repository :</td>
            <td><a href="https://github.com/HackYourFuture/alumni" target="_blank">alumni</a></td>
          </tr>
          <tr>
            <td class="label">Description :</td>
            <td>Alumni Guides, everything you need to land a job</td>
          </tr>
          <tr>
            <td class="label">Forks :</td>
            <td>9</td>
          </tr>
          <tr>
            <td class="label">Updated :</td>
            <td>6/21/2019, 1:38:21 PM</td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="right-div whiteframe">
      <p class="contributor-header">Contributions</p>
      <ul class="contributor-list">
        <li class="contributor-item" aria-label="gijscor" tabindex="0">
          <img
            src="https://avatars0.githubusercontent.com/u/15912395?v=4"
            height="48"
            class="contributor-avatar"
          />
          <div class="contributor-data">
            <div>gijscor</div>
            <div class="contributor-badge">47</div>
          </div>
        </li>
        <li class="contributor-item" aria-label="wouterkleijn" tabindex="0">
          <img
            src="https://avatars1.githubusercontent.com/u/37488847?v=4"
            height="48"
            class="contributor-avatar"
          />
          <div class="contributor-data">
            <div>wouterkleijn</div>
            <div class="contributor-badge">11</div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
```

Can you write JavaScript code, using `createAndAppend()` that creates something similar to what I have?